app/src/main/java/edu/mills/cs115/fruitthief/MainActivity.kt

```diff
@@ -0,0 +1,118 @@
+import androidx.navigation.findNavController
+import androidx.appcompat.app.AppCompatActivity
+import androidx.appcompat.widget.Toolbar
+import androidx.core.app.ActivityCompat
+import androidx.core.content.ContextCompat
+import androidx.databinding.DataBindingUtil
+import androidx.drawerlayout.widget.DrawerLayout
+import androidx.fragment.app.add
+import androidx.fragment.app.commit
+import androidx.navigation.NavController
+import androidx.navigation.NavDestination
+import androidx.navigation.ui.*
+import edu.mills.cs115.fruitthief.databinding.ActivityMainBinding
+import androidx.fragment.app.add
+import androidx.fragment.app.commit
+import com.google.android.gms.location.FusedLocationProviderClient
+import com.google.android.gms.location.LocationServices
+import com.google.android.gms.maps.GoogleMap
+import com.google.android.gms.maps.OnMapReadyCallback
+import com.google.android.gms.maps.SupportMapFragment
+import com.google.android.gms.maps.model.LatLng
+import com.google.android.gms.maps.model.MarkerOptions
+import edu.mills.cs115.fruitthief.database.FruitTreeDatabase
+import edu.mills.cs115.fruitthief.database.PopulateFruitTable
+import edu.mills.cs115.fruitthief.database.PopulateTreeTable
+import edu.mills.cs115.fruitthief.map.MapFragment
+import kotlinx.coroutines.runBlocking
+
+class MainActivity : AppCompatActivity() { //, OnMapReadyCallback {
+
+    private lateinit var binding: ActivityMainBinding
+    private lateinit var navController: NavController
+    private lateinit var drawerLayout: DrawerLayout
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)
+        navController = findNavController(R.id.nav_host_fragment)
+        drawerLayout = binding.drawerLayout
+        val application = requireNotNull(this).application
+        val dataSource = FruitTreeDatabase.getInstance(application).fruitTreeDAO
+
+        // clear tables for testing
+        runBlocking {
+            // clear for testing
+            dataSource.clearFruitTable()
+            dataSource.clearTrees()
+            PopulateFruitTable(dataSource)
+            PopulateTreeTable(dataSource)
+        }
+
+        // GPS Permission
+        if (ContextCompat.checkSelfPermission(this@MainActivity,
+                Manifest.permission.ACCESS_FINE_LOCATION) !==
+            PackageManager.PERMISSION_GRANTED) {
+            if (ActivityCompat.shouldShowRequestPermissionRationale(this@MainActivity,
+                    Manifest.permission.ACCESS_FINE_LOCATION)) {
+                ActivityCompat.requestPermissions(this@MainActivity,
+                    arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), 1)
+            } else {
+                ActivityCompat.requestPermissions(this@MainActivity,
+                    arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), 1)
```

Comment:
I don't see a difference between the if-body and else-body.

Reply:
I think the rationale is not the results of a true return, but to get it to run `shouldShowRequestPermissionRationale()` in the evaluation block if it's Android 11 or higher, but still request permission in Android 10 or below.

Reply:
That's a good point about the predicate having a side effect.

Reply:
Did you copy the code form somewhere? If so, I'd like to see the justification.

I was wrong about the predicate having a side effect.

In the absence of additional information, I'd drop the predicate and else-body and just have the if-body (and test it).

Reply:
If you just have the if body without the else it won't run on Android 10 or under because only Android 11 or higher has the function `ShowRequestPermissionRationale()` Android 10 and below can't pass that test. If there's an else statement it still allows it to run `ActivityCompat.requestPermissions()`. What I can do is remove that if statement completely as we're not providing extraneous information about why we need GPS data.  (It allows you to customize the permission access request window.)

Reply:
I just tested, it works without the if statement, and asks for permission.